### PR TITLE
expose reducers from SDK and consume in console

### PR DESCRIPTION
**Pull request by invincibleJai**
**Addresses:**

- [x] exposes reducers, actions from SDK
- [X] creates a package `host-app`  in SDK for host app developers
- [x] consume the exposed reducer and apply these in console

### CODE REVIEW

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx [hunk 1]
```diff
@@ -223,7 +223,7 @@ const CloudShellTerminal: React.FC<CloudShellTerminalProps &
 export const InternalCloudShellTerminal = CloudShellTerminal;
 
 const stateToProps = (state: RootState): StateProps => ({
-  user: state.UI.get('user'),
+  user: state.core.user,
```
Comment (andrewballantyne) [inline frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx:226]: I think we need to use selectors the SDK should provide here... going directly into our reducer structure is asking for trouble for apps. If we want properties to change, we'll never be able to do it if everyone writes their own walk through our state.

[Something like this](https://www.typescriptlang.org/play?#code/PTAEGUBEGkIFwIZwKagCoE8AOyDOBYAKDm1QGEB7AJ2QCVkATAVwGNkrxEVQBeUAbyKhQTXOwBcoXHCoBLAHYBzANxEAvqsJESOCDE7VUfQYWEtDkyjXrM2HLsk0aiLwiFABBLFgA2slkiyFPLwSKg6yNqknt4GNLwCQqAUcAAWEqDyTAC2AEbsTqAAZHrQcY6uRO4AogAeWBRiDKUQyD7ILHDUROby0lJtHV1UAGLUAKpiVAkAFNJhklBlw8gAlJaGNqzsnGEA2gDkouwHALq8AHyJpqA0cExUIfMoAHTmNC-HVE6arjW1CGyvlQogQikihF6-V68iG5V23GMyTSGQArAAaUDvZCSfgiKaSA4AKwYFGQB1AakpmihcHx7ASYnanWoYyok3YMxhcJWCLWmiAA)

File: frontend/packages/console-dynamic-plugin-sdk/README.md [hunk 2]
```diff
@@ -31,6 +31,7 @@ dynamic-demo-plugin/
 | `@openshift-console/dynamic-plugin-sdk` | Provides core APIs, types and utilities used by dynamic plugins at runtime. |
 | `@openshift-console/dynamic-plugin-sdk-webpack` | Provides webpack plugin `ConsoleRemotePlugin` used to build all dynamic plugin assets. |
 | `@openshift-console/dynamic-plugin-sdk-internal` | Internal package exposing additional code. |
+| `@openshift-console/dynamic-plugin-sdk-host-app` | Provides APIs, Component, and utilities for host applications i.e reducers, actions, etc |
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/README.md:34]: FYI @vojtechszocs

File: frontend/packages/console-dynamic-plugin-sdk/src/app/core/reducers/coreSelector.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+type GetImpersonate = (state: RootState) => { impersonate: ImpersonateKind };
+type GetUser = (state: RootState) => { user: UserKind };
+
+/**
+ * It provides impersonation details from the redux store.
+ * @param state the root state
+ * @return The the impersonate state.
+ * * */
+export const getImpersonate: GetImpersonate = (state) => {
+  return { impersonate: state.core.impersonate };
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/core/reducers/coreSelector.ts:12]: Oh hey, you have selectors! -- why are you not using them & why are they object responses?

File: frontend/packages/console-dynamic-plugin-sdk/src/api/host-app-api.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+export { default as coreReducer } from '@console/dynamic-plugin-sdk/src/app/core/reducers/core';
+export {
+  setUser,
+  beginImpersonate,
+  endImpersonate,
+} from '@console/dynamic-plugin-sdk/src/app/core/actions/core';
+export {
+  getUser,
+  getImpersonate,
+} from '@console/dynamic-plugin-sdk/src/app/core/reducers/coreSelector';
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/api/host-app-api.ts:11]: Isn't this backwards? Shouldn't our host-app-api be in here and exported via the other files?

File: frontend/public/actions/features.ts [hunk 5]
```diff
@@ -18,10 +18,11 @@ import store from '../redux';
 import { GroupModel, UserModel, VolumeSnapshotContentModel } from '../models';
 import { ClusterVersionKind } from '../module/k8s';
 import { receivedResources } from './k8s';
-import { setClusterID, setCreateProjectMessage, setUser } from './common';
+import { setClusterID, setCreateProjectMessage } from './common';
 import client, { fetchURL } from '../graphql/client';
 import { SSARQuery } from './features.gql';
 import { SSARQueryType, SSARQueryVariables } from '../../@types/console/generated/graphql-schema';
+import { setUser } from '@console/dynamic-plugin-sdk/src/app/core/actions/core';
```
Comment (andrewballantyne) [inline frontend/public/actions/features.ts:25]: This seems oddly specific... are you avoiding some circle reference issue? Not not just the SDK package?

File: frontend/public/actions/ui.ts [hunk 6]
```diff
@@ -14,9 +14,13 @@ import { K8sResourceKind, PodKind, NodeKind } from '../module/k8s';
...
 import { detectFeatures, clearSSARFlags } from './features';
 import { OverviewSpecialGroup } from '../components/overview/constants';
-import { setClusterID, setCreateProjectMessage, setUser } from './common';
+import { setClusterID, setCreateProjectMessage } from './common';
 import { Rule } from '../components/monitoring/types';
 import { subsClient } from '../graphql/client';
+import {
+  beginImpersonate,
+  endImpersonate,
+} from '@console/dynamic-plugin-sdk/src/app/core/actions/core';
```
Comment (andrewballantyne) [inline frontend/public/actions/ui.ts:23]: Same Q here

File: frontend/public/actions/ui.ts [hunk 7]
```diff
@@ -124,7 +125,7 @@ export const getNamespacedResources = () => {
 };
 
 export const getActiveNamespace = (): string => store.getState().UI.get('activeNamespace');
-export const getActiveUserName = (): string => store.getState().UI.get('user')?.metadata?.name;
+export const getActiveUserName = (): string => store.getState().core.user?.metadata?.name;
```
Comment (andrewballantyne) [inline frontend/public/actions/ui.ts:128]: I trust you'll find all of these and use the selector and I don't have to call them all out 🙂 ; even if these functions really should never exist.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Looking awesome, just a few comments.

File: frontend/public/actions/features.ts (hunk 5 above)
Comment (invincibleJai) [reply frontend/public/actions/features.ts:25]: make sense, it was an auto import. I doubt there would be a circular reference issue. will update

File: frontend/packages/console-dynamic-plugin-sdk/src/api/host-app-api.ts (hunk 4 above)
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/api/host-app-api.ts:11]: we are exporting as a new package here so, this should be its own export for host-app as we have for `core-api` , `internal-kubevirt` etc as in her https://github.com/openshift/console/tree/master/frontend/packages/console-dynamic-plugin-sdk/src/api

let me know if am missing something here.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/core/reducers/coreSelector.ts (hunk 3 above)
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/core/reducers/coreSelector.ts:12]: yeah can use them directly, make sense. Thanks :)

File: frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx (hunk 1 above)
Comment (invincibleJai) [reply frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx:226]: yeah make sense, I just update the current state with new key access. Have made use of selectors which have already been exposed.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/core/reducers/index.ts [hunk 8]
```diff
@@ -0,0 +1,2 @@
+export { coreReducer } from './core';
+export * from './coreSelector';
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/core/reducers/index.ts:2]:
```suggestion
export * from './coreSelectors';
```
Singular denotes one like a default export and some additional items... 

Just like it's called `utils` and not `util` when making some utility file.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/redux.ts [hunk 9]
```diff
@@ -1,5 +1,5 @@
-import coreReducer from './core/reducers/core';
+import { coreReducer } from './core/reducers/core';
 
-export const baseReducers = Object.freeze({
+export const SDKReducer = Object.freeze({
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/redux.ts:3]:
```suggestion
export const SDKReducers = Object.freeze({
```
When someone imports our reducers, we don't want them thinking that it's a singular one (or a combined set) that will take a single prop. This is our "list" of our internal reducers... they should just spread these forward and treat it opaque-like.

Maybe we should put a JS Doc comment here. Something like:

```
/**
 * Dynamic Plugin SDK Redux store reducers
 *
 * If you use Redux, these can be spread into the root of your store to provide an integrated SDK.
 * If you do not use Redux, these will be provided via the SDK Redux Store.
 */
export const SDKReducers = Object.freeze({
```

File: frontend/packages/console-dynamic-plugin-sdk/src/app/redux.ts [hunk 10]
```diff
@@ -1,5 +1,5 @@
...
   core: coreReducer,
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/app/redux.ts:4]: I think we'll need another story to deal with the naming of this... the more I think of it the more we either need to provide some global prefix context for App stores (like [basename](https://v5.reactrouter.com/web/api/BrowserRouter/basename-string) prefixes but for Redux), or like just prefix everything with SDK; `sdkCore` for instance. 

Today, any app that uses `core` in their store will lose it when they spread us into theirs.

File: frontend/public/components/app.jsx [hunk 11]
```diff
@@ -270,7 +271,7 @@ const CaptureTelemetry = React.memo(() => {
   const fireTelemetryEvent = useTelemetry();
 
   // notify of identity change
-  const user = useSelector(({ UI }) => UI.get('user'));
+  const user = useSelector(getUser) || {};
```
Comment (andrewballantyne) [inline frontend/public/components/app.jsx:274]: You added functionality here... any reason why the default?

File: frontend/public/components/edit-yaml.jsx [hunk 12]
```diff
@@ -47,10 +47,10 @@ const generateObjToLoad = (templateExtensions, kind, id, yaml, namespace = 'defa
   return sampleObj;
 };
 
-const stateToProps = ({ k8s, UI }) => ({
-  activeNamespace: UI.get('activeNamespace'),
-  impersonate: UI.get('impersonate'),
-  models: k8s.getIn(['RESOURCES', 'models']),
+const stateToProps = (state) => ({
+  activeNamespace: state.UI.get('activeNamespace'),
+  impersonate: getImpersonate(),
```
Comment (andrewballantyne) [inline frontend/public/components/edit-yaml.jsx:52]:
```suggestion
  impersonate: getImpersonate(state),
```
?

File: frontend/public/components/utils/kebab.tsx [hunk 13]
```diff
@@ -106,7 +107,7 @@ const KebabItem_: React.FC<KebabItemProps & { isAllowed: boolean }> = ({
     </button>
   );
 };
-const KebabItemAccessReview_ = (props: KebabItemProps & { impersonate: string }) => {
+const KebabItemAccessReview_ = (props: KebabItemProps & { impersonate: ImpersonateKind }) => {
```
Comment (andrewballantyne) [inline frontend/public/components/utils/kebab.tsx:110]: This is weird, but tracking this change I think it's right. `string` was the wrong type.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
I see one issue and a few suggestions. Looking very close to being good to go.

File: frontend/packages/console-dynamic-plugin-sdk/src/api/host-app-api.ts (hunk 4 above)
Comment (andrewballantyne) [reply frontend/packages/console-dynamic-plugin-sdk/src/api/host-app-api.ts]: You are quite right... apologizes. I thought of us making this where things were defined, but it makes more sense the way you have it. They are exposed where they need to be exposed, but written in a single spot.

File: frontend/public/components/utils/kebab.tsx (hunk 13 above)
Comment (invincibleJai) [reply frontend/public/components/utils/kebab.tsx:110]: yeah, it was wrong ... :)

File: frontend/public/components/edit-yaml.jsx (hunk 12 above)
Comment (invincibleJai) [reply frontend/public/components/edit-yaml.jsx:52]: ah, missed this.

File: frontend/packages/console-dynamic-plugin-sdk/src/app/redux.ts (hunk 10 above)
Comment (invincibleJai) [reply frontend/packages/console-dynamic-plugin-sdk/src/app/redux.ts:4]: yeah for now have made use of `sdkCore`.

File: frontend/public/components/app.jsx (hunk 11 above)
Comment (invincibleJai) [reply frontend/public/components/app.jsx:274]: not needed, updated.

**Review by andrewballantyne — APPROVED** [review]
/lgtm

We are not documenting the SDK at this time, adding the labels.
/label docs-approved
/label px-approved

**Comment by invincibleJai** [comment]
/cc @yapei  needed your help with QE verification.

As part of this PR have migrated some of the existing reducers in SDK so need to see if the overall app works and user impersonation happens as expected.

**Comment by yapei** [comment]
did a regression testing about dynamic plugins and user impersonate, working well
/label qe-approved

**Pull request merged**